src/NuGet.Core/NuGet.Build.Tasks/AssignReferencePropertiesTask.cs

```diff
@@ -0,0 +1,127 @@
+        /// </summary>
+        public ITaskItem[] AnnotatedProjectReferences { get; set; }
+
+        /// <summary>
+        /// The project references with assigned properties.
+        /// </summary>
+        [Output]
+        public ITaskItem[] AssignedProjects { get; set; }
+
+        public override bool Execute()
+        {
+            var log = new MSBuildLogger(Log);
+
+            BuildTasksUtility.LogInputParam(log, nameof(CurrentProjectTargetFramework), CurrentProjectTargetFramework);
+
+            if (AnnotatedProjectReferences != null)
+            {
+                log.LogDebug($"(in) AnnotatedProjectReferences '{string.Join(";", AnnotatedProjectReferences.Select(p => p.ItemSpec))}'");
+            }
+            else
+            {
+                return !Log.HasLoggedErrors;
+            }
+
+            var frameworkToMatch = NuGetFramework.Parse(CurrentProjectTargetFramework);
+            if (frameworkToMatch.IsUnsupported)
+            {
+                log.LogError(string.Format(Strings.UnsupportedTargetFramework, CurrentProjectTargetFramework));
+                return false;
+            }
+
+            AssignedProjects = AnnotatedProjectReferences.Select(project => AssignPropertiesForSingleReference(project, frameworkToMatch)).ToArray();
+
+            return !Log.HasLoggedErrors;
+        }
+
+        private ITaskItem AssignPropertiesForSingleReference(ITaskItem project, NuGetFramework currentProjectTargetFramework)
+        {
+            var itemWithProperties = new TaskItem(project);
+
+            var targetFramework = project.GetMetadata("TargetFrameworks");
+
+            if (string.IsNullOrEmpty(targetFramework))
+            {
+                // No target frameworks set, nothing to do.
+                return itemWithProperties;
+            }
+
+            var possibleTargetFrameworks = MSBuildStringUtility.Split(targetFramework);
+            var possibleNuGetFrameworks = possibleTargetFrameworks.Select(ParseFramework).ToList();
+            var nearestNuGetFramework = NuGetFrameworkUtility.GetNearest(possibleTargetFrameworks, currentProjectTargetFramework, NuGetFramework.Parse);
+
+            if (nearestNuGetFramework != null)
+            {
+                if (TryConvertItemMetadataToBool(project, "HasSingleTargetFramework", out var singleTargetFramework) && singleTargetFramework)
+                {
+                    // If the project has only one TF, don't specify it. It will go directly to the inner build anyway and we don't
+                    // want to redundantly specify a global property, which can cause a race.
+                    var undefineProperties = itemWithProperties.GetMetadata("UndefineProperties");
+
+                    itemWithProperties.SetMetadata("UndefineProperties",
+                        string.IsNullOrEmpty(undefineProperties)
+                            ? "TargetFramework"
+                            : undefineProperties + ";TargetFramework");
```

Comment:
IMHO, this is moving too much build logic in to NuGet. I was expecting the nuget task to be responsible for calculating nearest target framework and nothing else.

Reply:
I suspect some handling of Rid-agnostic is missing here, which adds to my concern that we're handling this sort of thing here in nuget.